src/CabalHelper/Runtime/Main.hs

```diff
@@ -554,14 +611,51 @@ removeInplaceDeps _v lbi pd clbi = let
 
 
 #if CH_MIN_VERSION_Cabal(2,0,0)
-recursiveIncludeDirs :: Map.Map UnitId ([UnitId],[FilePath], [(OpenUnitId, ModuleRenaming)])
-                     -> UnitId -> ([FilePath],[(OpenUnitId, ModuleRenaming)])
-recursiveIncludeDirs includeDirs unit = go ([],[]) [unit]
+recursiveDepInfo lbi v distdir = do
+  includeDirs <- componentsMap lbi v distdir $ \c clbi bi -> do
+    return (componentUnitId clbi
+           , ( componentInternalDeps clbi
+             , hsSourceDirs bi
+             , componentIncludes clbi
+             , componentEntrypoints c))
+  return $ Map.fromList $ map snd includeDirs
+
+type SubDeps = ([UnitId], [FilePath], [(OpenUnitId, ModuleRenaming)], ChEntrypoint)
+
+recursiveIncludeDirs :: Map.Map UnitId SubDeps
+                     -> UnitId -> ([FilePath], [(OpenUnitId, ModuleRenaming)]
+                                  , ChEntrypoint)
+recursiveIncludeDirs includeDirs unit = go ([],[],Nothing) [unit]
+  where
+    go (afp,aci,amep) [] = (afp,aci,maybe (error "recursiveIncludeDirs:no ChEntrypoint") id amep)
+    go acc@(afp,aci,amep) (u:us) = case Map.lookup u includeDirs of
+      Nothing -> go acc us
+      Just (us',sfp,sci,sep) -> go (afp++sfp,aci++sci,combineEp amep sep) (us++us')
+
+needsBuildOutput :: Map.Map UnitId SubDeps
+                     -> UnitId -> NeedsBuildOutput
+needsBuildOutput includeDirs unit = go [unit]
   where
-    go acc [] = acc
-    go (acp,acd) (u:us) = case Map.lookup u includeDirs of
-      Nothing -> go (acp,acd) us
-      Just (us',fps,pds) -> go (acp++fps,acd++pds) (us++us')
+    isIndef (IndefFullUnitId _ _) = True
+    isIndef _                     = False
+    go [] = NoBuildOutput
+    go (u:us) = case Map.lookup u includeDirs of
+      Nothing -> go us
+      Just (us',sfp,sci,sep) ->
+        if any (isIndef . fst) sci
+          then ProduceBuildOutput
+          else go (us++us')
+
+combineEp Nothing e = Just e
+combineEp (Just ChSetupEntrypoint) e = Just e
+combineEp (Just (ChLibEntrypoint es1 os1 ss1)) (ChLibEntrypoint es2 os2 ss2) = Just (ChLibEntrypoint (nub $ es2++es1) (nub $ os2++os1) (nub $ ss2++ss1))
+combineEp (Just (ChLibEntrypoint es1 os1 ss1)) (ChExeEntrypoint  mi os2)     = Just (ChExeEntrypoint mi (nub $ os2++es1++os1++ss1))
+combineEp (Just (ChExeEntrypoint   _ os1))     (ChLibEntrypoint es2 os2 ss2) = Just (ChLibEntrypoint es2 (nub $ os2++os1) ss2)
+combineEp (Just (ChExeEntrypoint   _ os1))     (ChExeEntrypoint  mi os2)     = Just (ChExeEntrypoint mi  (nub $ os2++os1))
```

Comment:
Not sure what to make of this function. Half of it should really be a `Semigroup` instance for `ChEntrypoint` and the other half (the `Maybe`) stuff is probably just `mplus` or some such.

Reply:
I agree it is messy, but have no good ideas on how to improve it.  Working code alternatives welcomed.

Reply:
Something like this:
```haskell
instance Semigroup ChEntrypoint where
  mappend (ChLibEntrypoint es1 os1 ss1) (ChLibEntrypoint es2 os2 ss2)
    = ChLibEntrypoint (nub $ es1 ++ es2) (nub $ os1 ++ os2) (nub $ ss1 ++ ss2)
  -- etc.

combineEp maybeEp1 ep2 = mappend ep2 <$> maybeEp1
```
Things I didn't check:

* [ ] Which `Semigroup` package to use
* [ ] Whether the order of the lists actually matters

I think the `Semigroup` instance can't be made much easier.

Reply:
Looking at this again I think a Semigroup instance just isn't appropriate here, it seems to me you're trying to use `ChEntrypoint` as a container for intermediate data for which it is just not fit for. Especially the conversions between `ChExe <-> ChLib` make this obvious and really don't fit what a Semigroup instance for this type should do in my mind (though it could probably be law abiding anyways). Previously I thought all the cases were just of the form `ChXEntrypoint, ChXEntrypoint -> ChXEntrypoint` which is why I suggested Semigroup.

Reply:
Does `recursiveIncludeDirs` really need to return a `ChXEntrypoint`? It didn't need to previously, and if it doesn't need to, the function could be rewritten in a much simpler way using some sort of `fold` I think. (Btw, in case of a successful `Map.lookup u includeDirs`, the function returns an "accumulated" `ChXEntrypoint` holding all the intermediate paths. Not sure whether this is intended.)